toolchain/check/dataflow_analysis.cpp

```diff
@@ -0,0 +1,345 @@
+      infos.push_back({val_bind->entity_name_id, binding_id});
+    }
+  }
+  return infos;
+}
+
+// This builds facts needed for carrying out dataflow analysis.
+static auto BuildDataflowFacts(const SemIR::File& sem_ir,
+                               SemIR::FunctionId function_id) -> DataflowFacts {
+  DataflowFacts facts;
+  const auto& function = sem_ir.functions().Get(function_id);
+
+  if (function.body_block_ids.empty()) {
+    return facts;
+  }
+
+  // Track ref params to treat assignments as uses.
+  Set<int32_t> ref_params;
+
+  // Collect definitions from parameters.
+  if (function.param_patterns_id.has_value()) {
+    auto param_patterns = sem_ir.inst_blocks().Get(function.param_patterns_id);
+    for (auto pattern_id : param_patterns) {
+      llvm::SmallVector<std::pair<SemIR::EntityNameId, SemIR::InstId>>
+          entity_names;
+      CollectEntityNamesFromPattern(sem_ir, pattern_id, entity_names);
+      for (auto [entity_name_id, def_inst_id] : entity_names) {
+        // Use the pattern_id as the instruction ID for the definition.
+        facts.defs.Insert(Fact{def_inst_id.index, entity_name_id.index});
+
+        // Identify ref parameters.
+        auto inst = sem_ir.insts().Get(pattern_id);
+        if (inst.Is<SemIR::RefParamPattern>()) {
+          ref_params.Insert(entity_name_id.index);
+        }
+      }
+    }
+  }
+
+  for (const auto& block_id : function.body_block_ids) {
+    const auto& block = sem_ir.inst_blocks().Get(block_id);
+
+    // Emit leader fact for non-empty blocks.
+    if (!block.empty()) {
+      facts.leaders.Insert(Fact{block_id.index, block.front().index});
+    }
+
+    // First pass: identify LHS of assignments to avoid counting them as uses.
+    Set<SemIR::InstId> assigned_lhs;
+    for (const auto& inst_id : block) {
+      auto inst = sem_ir.insts().Get(inst_id);
+      if (auto assign = inst.TryAs<SemIR::Assign>()) {
+        assigned_lhs.Insert(assign->lhs_id);
+      }
+    }
+
+    for (size_t i = 0; i < block.size(); ++i) {
+      auto inst_id = block[i];
+      auto inst = sem_ir.insts().Get(inst_id);
+
+      // Intra-block edge
+      if (i + 1 < block.size()) {
+        auto next_inst_id = block[i + 1];
+        facts.edges.Insert(Fact{inst_id.index, next_inst_id.index});
+      }
+
+      // 1. Definition (VarStorage)
+      if (inst.Is<SemIR::VarStorage>() || inst.Is<SemIR::ValueBinding>()) {
+        auto var_infos = GetVarInfos(sem_ir, inst_id);
+        for (auto [var_id, def_inst_id] : var_infos) {
+          facts.defs.Insert(Fact{def_inst_id.index, var_id.index});
+        }
+      }
+
+      // 2. Assignment
+      else if (auto assign = inst.TryAs<SemIR::Assign>()) {
+        auto var_infos = GetVarInfos(sem_ir, assign->lhs_id);
+        for (auto [var_id, _] : var_infos) {
+          facts.assigns.Insert(Fact{inst_id.index, var_id.index});
+        }
+      }
+
+      // 3. Use (NameRef)
+      else if (inst.Is<SemIR::NameRef>()) {
```

Comment:
Given the structure of what you're doing here, can you please look at CARBON_KIND_SWITCH? We added it in order to reduce `else if` structures. For example, see [call.cpp](https://github.com/carbon-language/carbon-lang/blob/7c1798d96dfe798eb799c3a0b69e84988a04619c/toolchain/check/call.cpp#L328) (but there are many other examples)

Similar applies elsewhere in this PR where you're doing repeated queries on the same instruction, such as handle_binding_pattern.cpp

Style note, I would typically expect the close brace on the same line (e.g. `} else`). For reference, https://google.github.io/styleguide/cppguide.html#Formatting_Looping_Branching